### ISSUE TITLE
S3 Synchronisation Error on Archlinux

### ISSUE THREAD

**Issue by allblue** [laurent22/joplin#4941]
## Environment

Joplin version:  1.7.11 or 1.8.3
Platform:  Archlinux
OS specifics: archlinux 5.12.1-arch1-1

## Steps to reproduce
The same S3 configuration runs normally in Windows and Android applications, but an error is reported in Archlinux

1. Install Joplin-desktop from AUR
2. S3 Compatible configuration 
3. Test s3 configuration error

## Logfile

```
2021-05-09 09:45:44: "Database was open successfully"
2021-05-09 09:45:44: "Checking for database schema update..."
2021-05-09 09:45:44: "Current database version", "{"version":34,"table_fields_version":34}"
2021-05-09 09:45:44: "Upgrading database from version 34"
2021-05-09 09:45:44: "New version: 34. Previously recorded version: 34"
2021-05-09 09:45:44: "KeychainService: checking if keychain supported"
2021-05-09 09:45:44: "KeychainService: check was already done - skipping. Supported:", "0"
2021-05-09 09:45:44: App: "Client ID: 62b48c6e34034159aed7fe0ecdd5f831"
2021-05-09 09:45:44: "Trying to load 0 master keys..."
2021-05-09 09:45:44: "Loaded master keys: 0"
2021-05-09 09:45:44: "ResourceFetcher: Auto-add resources: Mode: always"
2021-05-09 09:45:44: "ResourceFetcher: Auto-added resources: 0"
2021-05-09 09:45:44: "Trying to load 0 master keys..."
2021-05-09 09:45:44: "Loaded master keys: 0"
2021-05-09 09:45:44: "Trying to load 0 master keys..."
2021-05-09 09:45:44: "Loaded master keys: 0"
2021-05-09 09:45:44: "Trying to load 0 master keys..."
2021-05-09 09:45:44: "Loaded master keys: 0"
2021-05-09 09:45:44: "app.start: doing regular boot"
2021-05-09 09:45:44: "Trying to load 0 master keys..."
2021-05-09 09:45:44: "Loaded master keys: 0"
2021-05-09 09:45:44: "Trying to load 0 master keys..."
2021-05-09 09:45:44: "Loaded master keys: 0"
2021-05-09 09:45:44: "Trying to load 0 master keys..."
2021-05-09 09:45:44: "Loaded master keys: 0"
2021-05-09 09:45:44: "RevisionService::runInBackground: Starting background service with revision collection interval 600000"
2021-05-09 09:45:44: SpellCheckerServiceDriverNative: "Set effective language from "en-US" to "en-US""
2021-05-09 09:45:46: "Preparing scheduled sync"
2021-05-09 09:45:46: "DecryptionWorker: cannot start because no master key is currently loaded."
2021-05-09 09:45:46: "Starting scheduled sync"
2021-05-09 09:45:46: "Error: Do not have proper permissions to Bucket
Code: rejectedByTarget
Error: Do not have proper permissions to Bucket
    at FileApiDriverAmazonS3.get (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api-driver-amazon-s3.js:247:11)
    at async tryAndRepeat (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api.js:40:19)"
2021-05-09 09:45:46: "Operations completed: "
2021-05-09 09:45:46: "Total folders: 0"
2021-05-09 09:45:46: "Total notes: 0"
2021-05-09 09:45:46: "Total resources: 3"
2021-05-09 09:45:46: "There was some errors:"
2021-05-09 09:45:46: "Error: Do not have proper permissions to Bucket
Code: rejectedByTarget
Error: Do not have proper permissions to Bucket
    at FileApiDriverAmazonS3.get (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api-driver-amazon-s3.js:247:11)
    at async tryAndRepeat (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api.js:40:19)"
2021-05-09 09:45:46: "Updating all notifications..."
2021-05-09 09:45:46: "Garbage collecting alarms..."
2021-05-09 09:45:46: "Trying to reload note that has been deleted:", "null"
2021-05-09 09:45:47: "DecryptionWorker: cannot start because no master key is currently loaded."
2021-05-09 09:45:48: "RevisionService::maintenance: Starting..."
2021-05-09 09:45:48: "RevisionService::maintenance: Service is enabled"
2021-05-09 09:45:48: "RevisionService::collectRevisions: Created revisions for 0 notes"
2021-05-09 09:45:48: "RevisionService::maintenance: Done in 129ms"
2021-05-09 09:45:54: "SearchEngine: Updating FTS table..."
2021-05-09 09:45:54: "SearchEngine: Updated FTS table in 1ms. Inserted: 0. Deleted: 0"
2021-05-09 09:45:55: "Preparing scheduled sync"
2021-05-09 09:45:55: "Starting scheduled sync"
2021-05-09 09:45:55: "Error: Do not have proper permissions to Bucket
Code: rejectedByTarget
Error: Do not have proper permissions to Bucket
    at FileApiDriverAmazonS3.get (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api-driver-amazon-s3.js:247:11)
    at async tryAndRepeat (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api.js:40:19)"
2021-05-09 09:45:55: "Operations completed: "
2021-05-09 09:45:55: "Total folders: 0"
2021-05-09 09:45:55: "Total notes: 0"
2021-05-09 09:45:55: "Total resources: 3"
2021-05-09 09:45:55: "There was some errors:"
2021-05-09 09:45:55: "Error: Do not have proper permissions to Bucket
Code: rejectedByTarget
Error: Do not have proper permissions to Bucket
    at FileApiDriverAmazonS3.get (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api-driver-amazon-s3.js:247:11)
    at async tryAndRepeat (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api.js:40:19)"
2021-05-09 09:45:56: "Trying to reload note that has been deleted:", "null"
2021-05-09 09:45:57: "DecryptionWorker: cannot start because no master key is currently loaded."
2021-05-09 09:46:14: "ResourceService::indexNoteResources: Start"
2021-05-09 09:46:14: "ResourceService::indexNoteResources: Completed"
2021-05-09 09:46:14: "ResourceService::deleteOrphanResources:", "[]"
2021-05-09 09:50:56: "Running background sync on timer..."
2021-05-09 09:50:56: "Preparing scheduled sync"
2021-05-09 09:50:56: "Starting scheduled sync"
2021-05-09 09:50:56: "Error: Do not have proper permissions to Bucket
Code: rejectedByTarget
Error: Do not have proper permissions to Bucket
    at FileApiDriverAmazonS3.get (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api-driver-amazon-s3.js:247:11)
    at async tryAndRepeat (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api.js:40:19)"
2021-05-09 09:50:56: "Operations completed: "
2021-05-09 09:50:56: "Total folders: 0"
2021-05-09 09:50:56: "Total notes: 0"
2021-05-09 09:50:56: "Total resources: 3"
2021-05-09 09:50:56: "There was some errors:"
2021-05-09 09:50:56: "Error: Do not have proper permissions to Bucket
Code: rejectedByTarget
Error: Do not have proper permissions to Bucket
    at FileApiDriverAmazonS3.get (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api-driver-amazon-s3.js:247:11)
    at async tryAndRepeat (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api.js:40:19)"
2021-05-09 09:50:56: "Trying to reload note that has been deleted:", "null"
2021-05-09 09:50:57: "DecryptionWorker: cannot start because no master key is currently loaded."
```

**Comment by laukaichung** [comment]
The document doesn't include all of the required permissions. I need to add `s3:GetObject` and `s3:GetObjectVersion` for it to work. This is my IAM policy:

{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [
                "s3:PutObject",
                "s3:DeleteObjectVersion",
                "s3:ListBucket",
                "s3:DeleteObject",
                "s3:GetBucketLocation",
                "s3:GetObjectVersion"
            ],
            "Resource": [
                "arn:aws:s3:::xxx/*",
                "arn:aws:s3:::xxx"
            ]
        }
    ]
}

After updating the policy, wait a few seconds for s3 to reload and click `sync` to test.

**Comment by allblue** [comment]
Excuse me, where is this document? 

>  
> The document doesn't include all of the required permissions. I need to add `s3:GetObject` and `s3:GetObjectVersion` for it to work. This is my IAM policy:
> 
> ```
> {
>     "Version": "2012-10-17",
>     "Statement": [
>         {
>             "Sid": "VisualEditor0",
>             "Effect": "Allow",
>             "Action": [
>                 "s3:PutObject",
>                 "s3:DeleteObjectVersion",
>                 "s3:ListBucket",
>                 "s3:DeleteObject",
>                 "s3:GetObject",
>                 "s3:GetBucketLocation",
>                 "s3:GetObjectVersion"
>             ],
>             "Resource": [
>                 "arn:aws:s3:::xxx/*",
>                 "arn:aws:s3:::xxx"
>             ]
>         }
>     ]
> }
> ```
> 
> After updating the policy, wait a few seconds for s3 to reload and click `sync` to test.
> Also, I have to restart the app after changing the credentials.

**Comment by JackGruber** [comment]
On the start side of the Joplin Homepage.

https://joplinapp.org/#aws-s3-synchronisation

**Comment by allblue** [comment]
I think it’s not the s3 configuration problem, because it works normally on both windows and android, including additions, deletions, and changes. 

> 
> 
> On the start side of the Joplin Homepage.
> 
> https://joplinapp.org/#aws-s3-synchronisation

**Comment by leematos** [comment]
I just spent some time in the joplin S3 code so chiming in here. It's strange because the file driver api is the _same_ across all platforms but I did see iOS errors pop up that haven't hit other platforms so it's possible.

Your error message is interesting:

2021-05-09 09:45:46: "Error: Do not have proper permissions to Bucket
Code: rejectedByTarget
Error: Do not have proper permissions to Bucket
    at FileApiDriverAmazonS3.get (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api-driver-amazon-s3.js:247:11)
    at async tryAndRepeat (/usr/share/joplin-desktop/resources/app.asar/node_modules/@joplin/lib/file-api.js:40:19)"

---

1. Who is your bucket provider?
2. Are you using the same keys on all 3 hosts?
3. This error is attempting a Get which https://github.com/laurent22/joplin/issues/4941#issuecomment-844709716 talked about. Does your archlinux key have get permissions? If it does, you might want to reach out to your S3 provider to troubleshoot with them. 

This is the error you are getting: https://github.com/laurent22/joplin/blob/0765cf5955ffb254020ca2d1cec53fdeafda0132/packages/lib/file-api-driver-amazon-s3.js#L246-L247

Which means that your provider said that your keys don't have access to `get` the bucket.
